### ISSUE TITLE
Remove `rev` import from `prim/mat/fun/gp_dot_prod_cov.hpp`

### ISSUE THREAD

**Issue by ChrisChiasson** [stan-dev/math#887]
https://github.com/stan-dev/math/blob/c6a8d7f9f758ed42b2ee27eed230962304475783/stan/math/prim/mat/fun/gp_dot_prod_cov.hpp#L11

Is this line doing what I think it is doing? I am getting an error in my tests where some (but not all) reverse mode functionality appears to be instantiated in a test dedicated entirely to forward mode.

Should it be pointing here instead?
https://github.com/stan-dev/math/blob/develop/stan/math/prim/mat/fun/dot_product.hpp

**Comment by ChrisChiasson** [comment]
This was the test. Changing `rev` to `prim` made the problem go away.
````
clang++-3.8 -Wall -I . -isystem lib/eigen_3.3.3 -isystem lib/boost_1.66.0 -isystem lib/sundials_3.1.0/include -std=c++1y -DBOOST_RESULT_OF_USE_TR1 -DBOOST_NO_DECLTYPE -DBOOST_DISABLE_ASSERTS -DBOOST_PHOENIX_NO_VARIADIC_EXPRESSION -Wno-unused-function -Wno-uninitialized -stdlib=libc++ -DGTEST_USE_OWN_TR1_TUPLE -DGTEST_HAS_PTHREAD=0 -isystem lib/gtest_1.7.0/include -isystem lib/gtest_1.7.0 -O3 -DGTEST_USE_OWN_TR1_TUPLE -DGTEST_HAS_PTHREAD=0 -isystem lib/gtest_1.7.0/include -isystem lib/gtest_1.7.0 -O3 -DNO_FPRINTF_OUTPUT -pipe  -c -o test/unit/math/fwd/mat/fun/accumulator_test.o test/unit/math/fwd/mat/fun/accumulator_test.cpp
````

**Comment by syclik** [comment]
Just based on the title, please don't include anything in the rev folder from prim. 

More specifically, the right thing to do is to include the prim header. Then use the appropriate top-level include. This works because the meta programs are instantiated in the correct way (which is very specific when it comes to instantiating against boost and eigen). 

Mind closing the issue?

**Comment by ChrisChiasson** [comment]
(i am saying that the source code is _currently_ including rev from prim)

**Comment by syclik** [comment]
Yes. Absolutely! Will get that fixed. 

Thank you for being so polite about my misunderstanding and clarifying for me. Much appreciated.

**Comment by ChrisChiasson** [comment]
We all have our short fuse days. I was banging my desk near the end of the
day and dropping all kinds of expletives in my car because, let's face it,
coding drives people insane.

**Comment by bob-carpenter** [comment]
Thanks for hanging in there.  @syclik and I went through some seriously rough patches before version 1 of Stan was released where it'd take two days of solid work just to get back to something that would compile.  We're better at it now, but it was a very slow and incredibly frustrating learning process.

C++ is particularly nasty because it leaves some behavior undefined and it relies on template metaprogramming which has no decent debugging tools. Hence the memes:

https://memegenerator.net/instance/39604198/grumpy-cat-i-love-cc-it-make-people-cry-

**Comment by ChrisChiasson** [comment]
Commit 8e02dc780ace6137993ac15ddc81e00f6685b295 from a month ago brings in some unit tests to https://github.com/stan-dev/math/blame/develop/test/unit/math/prim/mat/fun/add_diag_test.cpp that reference var (rev) from prim as well. These have been pulled into my complex math build upon rebasing (just like the original issue above) and are causing my tests to fail (as was the original issue above until I fixed it) in https://github.com/stan-dev/math/pull/789

**Comment by ChrisChiasson** [comment]
Turns out they can be switched to use double, so that's what I did

**Comment by bob-carpenter** [comment]
> I am going to comment them out for now until someone tells me that isn't ok

Thanks for asking.  We don't want commented-out code in Stan.  If it's not being used, remove it.  Otherwise, it's a big distraction to anyone reading the code (who will not only wonder why it's commented out but what it originally did and how it corresponds to the included code).  Teadability is one of our main goals for code because we have a big team of developers doing long term maintenance.

**Comment by ChrisChiasson** [comment]
Got it, BTW I was able to save the tests by converting them to double

**Comment by bob-carpenter** [comment]
rev shouldn't be included in prim.

**Comment by ChrisChiasson** [comment]
Not sure why issue is closed... ??? Currently rev is included in prim in the areas i called out. Issue is about preventing that from happening ???

**Comment by bob-carpenter** [comment]
I must have misunderstood what's going on.   To help everyone in the future, please try to formulate issues with a description of the problem or desired feature and the technical steps required.  This is formulated more as a question.

If there's an issue with the `develop` branch that `prim` is importing something from `rev` somewhere, please create a new issue to address that.  It should be titled something like "remove rev include from prim`.

Thanks!

**Comment by bob-carpenter** [comment]
P.S.  For actual discussions, we have discourse.

**Comment by seantalts** [comment]
Reopening - I think the title threw everyone off. I renamed it to "Remove `rev` import from `prim/mat/fun/gp_dot_prod_cov.hpp`", @ChrisChiasson please correct me if that doesn't represent the issue here.

**Comment by seantalts** [comment]
@ChrisChiasson do you have a PR for this by any chance? Or should I make one?

**Comment by ChrisChiasson** [comment]
the thread opening issue and the one in the comments are fixed in the complex math PR (it doesn't work without those fixes)

**Comment by drezap** [comment]
yeah I just came across this convo. This is my "wrong doing", I'll make a PR and remove this include, if it hasn't been done already. I implemented this, and it shouldn't be there.

**Comment by drezap** [comment]
changing it to a prim include

**Comment by ChrisChiasson** [comment]
Fyi (regarding testing server burden) this is fixed in the complex math
patch which is passing and under code review now

**Comment by drezap** [comment]
good to hear, I just opened it up for the "non-complex" math and we can close this issue once both of those are merged. Thanks @ChrisChiasson

**Comment by andrjohns** [comment]
Closed via #965
